### PR TITLE
Job tests

### PR DESCRIPTION
**Pull request by uanabitarte**
This PR includes the client tests of the PR #552 and the job tests

### CODE REVIEW

File: fixtures/jobs.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  job_to_update_{1..15}:
+    backupLocation: '@backupLocation'
+    client: '@client_10'
+    isActive: true
+    minNotificationLevel: 400
+    name: 'job_to_update_<current()>'
+    notificationsTo: ['owner']
+    path: '/some/fake/path'
+    policy: '@policy1'
+    useLocalPermissions: true
```
Comment (igorbga) [inline fixtures/jobs.yaml:21]: The same as my coment about clients.yaml

File: fixtures/scripts.yaml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      __construct: ['/var/spool/elkarbackup/uploads']
+      description: <text()>
+      name: 'script_not_job_pre'
+      scriptFile: <upload('/home/uanabitarte/script.sh')>
+      isClientPre: <boolean()>
+      isJobPre: false
+      isClientPost: <boolean()>
+      isJobPost: <boolean()>
+      
+
```
Comment (igorbga) [inline fixtures/scripts.yaml:82]: It says that there is no new line in the last lien, so I guess there is some empty space or something like that.

File: tests/api/ClientTest.php [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    }
+
+    public function testCreateClientRepeatedName(): void
+    {
+        $httpClient = $this->authenticate();
+        $client = ClientMother::base();
+        $this->postClient($httpClient, $client);
+        $this->postClient($httpClient, $client);
+        $this->assertResponseStatusCodeSame(400);
+        $this->assertHydraError("An exception occurred while executing 'INSERT INTO Client (description, isActive, name, url, quota, sshArgs, rsyncShortArgs, rsyncLongArgs, state, maxParallelJobs, data, owner_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [null, 1, \"".$client->getName()."\", \"\", -1, null, null, null, \"NOT READY\", 1, null, 1]:\n\nSQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '".$client->getName()."' for key 'Client.UNIQ_C0E801635E237E06'");
```
Comment (igorbga) [inline tests/api/ClientTest.php:68]: Is such a great deal to add a comment pointing to the already opened issue?
I think is the third review where I ask for it.

File: tests/api/ClientTest.php [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    {
+        $httpClient = $this->authenticate();
+        $scriptId = $this->getScriptId($httpClient, 'script_not_client_pre');
+        $client = ClientMother::withPreScripts([$scriptId]);
+        $this->postClient($httpClient, $client);
+        $this->assertResponseStatusCodeSame(400);
+        $this->assertResponseHeaderSame('content-type', 'application/ld+json; charset=utf-8');
+        $this->assertHydraError('Script "'.$scriptId.'" is not a client pre script');
+    }
+    public function testCreateClientNonExistentPostScript(): void
```
Comment (igorbga) [inline tests/api/ClientTest.php:102]: blank like missing

File: src/Api/DataTransformers/ClientOutputDataTransformer.php [hunk 5]
```diff
@@ -53,7 +53,7 @@ public function transform($data, string $to, array $context = [])
         $output->setId($data->getId());
         $output->setName($data->getName());
         $output->setUrl($data->getUrl());
-        $output->setQuota($data->getQuota());
+        $output->setQuota((int)$data->getQuota());
```
Comment (igorbga) [inline src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: Did you check by creating a client with a huge quota through the api?

File: fixtures/clients.yaml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+      owner: '@user'
+      maxParallelJobs: 1
+      
+    client_to_delete:
+      name: 'client_to_delete'
+      description: <text()>
+      isActive: <boolean()>
+      quota: -1
+      owner: '@user'
+      maxParallelJobs: 1
```
Comment (igorbga) [inline fixtures/clients.yaml:16]: client_to_delete ? Will it be deleted during a test?

Haven't we agreed to try to make tests as autonomous and self sustained as possible ?
Haven't we agreed not to fixturize any content that would be changed in the tests?

I know that right now we are resetting all our DB and fixtures for each execution of the DB but once our test cases increase we might want to separate the resetting of the database and the execution of them because the developer might just want to execute a single test or a single suite repeatedly while developing them. Well in fact I'm sure you have already experienced the inconvenience of having to wait for the DB to recreate just to check one test.

Please review all the cases where you are changing preloaded content (deletes and updates).

File: tests/api/ClientTest.php [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+        $client = $this->postClient($httpClient, $client);
+        $postScript = $this->getScriptId($httpClient, 'script_client_post');
+        $preScript = $this->getScriptId($httpClient, 'script_client_pre');
+        $updateClient = ClientMother::withAllParameters(
+            1,
+            "description updated",
+            true,
+            5,
+            [$postScript],
+            [$preScript],
```
Comment (igorbga) [inline tests/api/ClientTest.php:204]: More than one script per array in both cases please to at least make sure that the handling of the array is correct.

File: tests/api/JobTest.php [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+            "exclude pattern",
+            "include pattern",
+            true,
+            400,
+            "example@example.com",
+            ["owner", "email"],
+            "/some/random/path",
+            1,
+            [$postScript],
+            [$preScript],
```
Comment (igorbga) [inline tests/api/JobTest.php:39]: More than one script per array

File: tests/api/JobTest.php [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+        $this->assertResponseHeaderSame('content-type', 'application/ld+json; charset=utf-8');
+        $this->assertHydraError('Script "'.$scriptId.'" does not exist');
+    }
+
+    public function testDeleteJob(): void
+    {
+        $httpClient = $this->authenticate();
+        $iri = $this->findIriBy(Job::class, [
+            'name' => 'job_to_delete'
+        ]);
```
Comment (igorbga) [inline tests/api/JobTest.php:165]: Do not delete preloaded

**Review by igorbga — CHANGES_REQUESTED** [review]
Is good to see that once bases settled for ClientTests you have done JobTests quite fast.

Please pay attention to each comment and tell me whenever each one is resolved. Most of them were already posted in previous revious and since I only got feedback about a couple of them I thought that you might already have them resolved for this PR.

The most worrying point for me is the part where you use fixturized data to delete and update.

ObjectMothers look reasonable too but I see that you have a lot of methods that just change one field and you have the whole JSON inlined in each one. Why not have the base JSON as a constant (or call "base" static method) and then just adapt it. That way the mother's will be more concise and if you decide that the "default" value for something is better fit with another value you only have to change it once.

**Comment by igorbga** [comment]
Don't forget to merge master so that conflicting composer.lock is resolved.

File: tests/api/ClientTest.php (hunk 3 above)
Comment (uanabitarte) [reply tests/api/ClientTest.php:68]: I'm sorry, I thought you meant to add the comment in the PR conversation, not inside the code. I'm on it.

File: src/Api/DataTransformers/ClientOutputDataTransformer.php (hunk 5 above)
Comment (uanabitarte) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: I've been testing and data is lost any ways. For example, I have created a client with quota 9223372034539499520, and with or without the casting it returns the value 9223372034539500000, the only difference are the quotes. 

Also, I was not aware that when you set a quota from the UI, the value is converted and saved to bytes. So I should change that for the api too.

Comment (igorbga) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: Is the issue about losing data also present in the Web UI or is it just the API ? If it happens in both cases then just open a new issue so that it is solved.

About converting to bytes. What is the unit that the Web UI uses ? Ks? Mbs? Is always fixed or the user can change it? If the user can change it at will then we should just use Bytes in the API, make them work and be happy. If it is hardcoded to one unit then yes, let's stick to that in the API so that we don't get inconsistencies. Let's see what @elacunza thinks about this.

Comment (uanabitarte) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: Its just in the API.
It uses GBs, and it's always fixed. What's more, I think there is a bug because each time I edit and save the client's form, the form tries to convert the quota to Bytes (also when it's already converted), so eventually it will exceed the accepted value range.

**Comment by elacunza** [comment]
Just make sure that the number we set via API is was we get when reading back, then move on.

There's some mess with this, starting with using bytes as unit for quota, but it isn't our concern right now.

File: src/Api/DataTransformers/ClientOutputDataTransformer.php (hunk 5 above)
Comment (elacunza) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: So make API consistent, file a issue for the form. ;)

File: fixtures/clients.yaml (hunk 6 above)
Comment (uanabitarte) [reply fixtures/clients.yaml:16]: Ok. And in the cases that the request should fail and it should not change the content, should I use the preloaded content, or it is not convenient either?

File: src/Api/DataTransformers/ClientOutputDataTransformer.php (hunk 5 above)
Comment (uanabitarte) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: If the quota value given is larger than the max_safe_int (9007199254740991), json will round the value, so the value given and the one that it saves will not be consistent. Do I handle it to throw an Exception if the quota value is higher?

Comment (elacunza) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: Check that it is not greater than the MAX_SAFE_INT, if it is then return an error ("Out of bounds" or something like that). That limit seems reasonable for know (petabytes of quota).

Also, it must be clear the unit for that field in the API.

Comment (igorbga) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: If [comment in the code](https://github.com/elkarbackup/elkarbackup/blob/7d7459fd65e1e18b15f8243a520bf7a556bf7ba2/src/Form/Type/ClientType.php#L28) is right, the actual persisted value is in KBytes, not Bytes.

[When saving it is multiplied twice by 1024](https://github.com/elkarbackup/elkarbackup/blob/7d7459fd65e1e18b15f8243a520bf7a556bf7ba2/src/Form/Type/ClientType.php#L33) and [when presenting divided twice](https://github.com/elkarbackup/elkarbackup/blob/7d7459fd65e1e18b15f8243a520bf7a556bf7ba2/templates/default/fields.html.twig#L73).

Quite a mess but if the actual UI does not better either let's just be practical and don't lose more time with this.

In the API we return the quota in GBs and we also expect them in GB. For that max_safe_int in json is more than enough. "int" should suffix in your API classes, no need to send or receive the number as a string.

The problem might arise when multiplying the received value twice by 1024 to save it in the entity as an int, but in a 64bit system the max value for PHP is 9,223,372,036,854,775,807, so we probably will be safe unless we go well over 8,000,000,000,000 Gbyte, The UI doesn't seem to pay attention to this either, so that will be good enough for now.

To sum up:
- No fancy "string" number in the JSON just a regular int.
- Multiple and divide when going from Doctrine Entity <=> API Dto
- Make sure that for a reasonable high value (for example 123,456 Gbytes) it works as expected

This test will probably fail in a 32bit system, but that's on purpose, it seems the UI has the same issue and I'd rather tests fail and surface the problem when run in a 32bit system than let the user think that everything is ok.

Comment (elacunza) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: If there's no compelling reason to use GB in API, I'll prefer to use MiB, so that quotas smaller than 1GB can be configured.

Comment (uanabitarte) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: Ok, so then MiB for the API (receive and return) and save it in KB?

Comment (elacunza) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: Save it the same as does the WUI.

Comment (uanabitarte) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: Ok, I will make that changes in other branch as it is not a test related issue

Comment (igorbga) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: @elacunza keep in mind that if you use MB in the API and then view or update the client in the GUI you will probably lose that detail.

Comment (elacunza) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: Sorry, what is using WUI? I thought it was using MiB? (because of 1024*1024 multiplication...)

File: fixtures/clients.yaml (hunk 6 above)
Comment (igorbga) [reply fixtures/clients.yaml]: If it is not a big deal I think is safer to not use it either in that case. If there is a mistake and it does change it we might end up with misleading results from other tests that also used the fixtured object.

File: src/Api/DataTransformers/ClientOutputDataTransformer.php (hunk 5 above)
Comment (uanabitarte) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: Changes made in PR #560

Comment (igorbga) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php:56]: @uanabitarte PR #560 is merged, merge it in this branch.
Do tests still need to be changed or they already work as commented?
Once this is confirmed I can resolve this conversation and Eneko will be able to merge it.

Comment (uanabitarte) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php]: I think the only issue left is that the tests do not check if there is a well known client or job in the collection (GET /clients or /jobs). I could not find a way to assert the content of the array inside the json.
I.e., the response is a json and the elements are in the 'hydra:member' field, that is an array of jsons, so it should check whether one of those json matches with the well known element. 
There is a assertArraySubset function, but as I have read, it only works for associative arrays.

Comment (elacunza) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php]: I think you should be able to convert to PHP object/array any JSON string with json_decode() . You can then check array's values with in_array for example.

Comment (igorbga) [reply src/Api/DataTransformers/ClientOutputDataTransformer.php]: @uanabitarte checking that the result holds the expected data is important, so please try @elacunza's suggested approach.

If I'm not getting this wrong this is not related with the quota issue but I'll just unresolve this conversation to not generate more noise, once solved pleas resolve the conversation and I will approve.

**Review by igorbga — COMMENTED** [review]
To assert the content itself you do $client->getCompleteArray() for clients but $job->getData() . Why this distinction in method name? Aren't we supposedly checking the same in both cases?

**Comment by uanabitarte** [comment]
Because getData returns just the fields of the client entity and in getCompleteArray it is also returned the iri, id and type. We use getData, for example, when we try to create a client with the same attributes, in that case we do not want to send the iri with the data, it causes errors.

**Review by igorbga — APPROVED** [review]
